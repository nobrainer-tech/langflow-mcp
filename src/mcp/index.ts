#!/usr/bin/env node

import { LangflowMCPServer } from './server';
import { logger } from '../utils/logger';
import * as dotenv from 'dotenv';

dotenv.config();

process.on('uncaughtException', (error) => {
  if (process.env.MCP_MODE !== 'stdio') {
    console.error('Uncaught Exception:', error);
  }
  logger.error('Uncaught Exception:', error);
  process.exit(1);
});

process.on('unhandledRejection', (reason, promise) => {
  if (process.env.MCP_MODE !== 'stdio') {
    console.error('Unhandled Rejection at:', promise, 'reason:', reason);
  }
  logger.error('Unhandled Rejection:', reason);
  process.exit(1);
});

async function main() {
  try {
    const mode = process.env.MCP_MODE || 'stdio';

    if (mode === 'http') {
      logger.error('HTTP mode not yet implemented. Use stdio mode.');
      process.exit(1);
    }

    const server = new LangflowMCPServer();

    let isShuttingDown = false;
    const shutdown = async (signal: string = 'UNKNOWN') => {
      if (isShuttingDown) return;
      isShuttingDown = true;

      try {
        logger.info(`Shutdown initiated by: ${signal}`);
        await server.shutdown();

        if (process.stdin && !process.stdin.destroyed) {
          process.stdin.pause();
          process.stdin.destroy();
        }

        setTimeout(() => {
          logger.warn('Shutdown timeout exceeded, forcing exit');
          process.exit(0);
        }, 1000).unref();
      } catch (error) {
        logger.error('Error during shutdown:', error);
        process.exit(1);
      }
    };

    process.on('SIGTERM', () => shutdown('SIGTERM'));
    process.on('SIGINT', () => shutdown('SIGINT'));
    process.on('SIGHUP', () => shutdown('SIGHUP'));

    if (process.stdin.readable && !process.stdin.destroyed) {
      try {
        process.stdin.on('end', () => shutdown('STDIN_END'));
        process.stdin.on('close', () => shutdown('STDIN_CLOSE'));
      } catch (error) {
        logger.error('Failed to register stdin handlers:', error);
      }
    }

    await server.run();
  } catch (error) {
    logger.error('Failed to start MCP server:', error);
    process.exit(1);
  }
}

if (require.main === module) {
  main().catch(console.error);
}
