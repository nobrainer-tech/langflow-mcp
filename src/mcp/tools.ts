import { ToolDefinition } from '../types';

export const langflowMCPTools: ToolDefinition[] = [
  {
    name: 'create_flow',
    description: 'Create a new Langflow flow. Pass name and optional description, data, folder_id.',
    inputSchema: {
      type: 'object',
      properties: {
        name: {
          type: 'string',
          description: 'The name of the flow'
        },
        description: {
          type: 'string',
          description: 'Optional description of the flow'
        },
        data: {
          type: 'object',
          description: 'Optional flow data/configuration'
        },
        folder_id: {
          type: 'string',
          description: 'Optional folder ID to organize the flow'
        }
      },
      required: ['name']
    }
  },
  {
    name: 'list_flows',
    description: 'List all Langflow flows with optional pagination and filtering.',
    inputSchema: {
      type: 'object',
      properties: {
        page: {
          type: 'number',
          description: 'Page number for pagination (default: 1)'
        },
        size: {
          type: 'number',
          description: 'Number of items per page (default: 50, max: 100)'
        },
        folder_id: {
          type: 'string',
          description: 'Filter by folder ID'
        },
        components_only: {
          type: 'boolean',
          description: 'Return only components (default: false)'
        },
        get_all: {
          type: 'boolean',
          description: 'Get all flows without pagination (default: true)'
        }
      }
    }
  },
  {
    name: 'get_flow',
    description: 'Get details of a specific flow by ID.',
    inputSchema: {
      type: 'object',
      properties: {
        flow_id: {
          type: 'string',
          description: 'The ID of the flow to retrieve'
        }
      },
      required: ['flow_id']
    }
  },
  {
    name: 'update_flow',
    description: 'Update an existing flow. Pass flow_id and fields to update.',
    inputSchema: {
      type: 'object',
      properties: {
        flow_id: {
          type: 'string',
          description: 'The ID of the flow to update'
        },
        name: {
          type: 'string',
          description: 'New name for the flow'
        },
        description: {
          type: 'string',
          description: 'New description for the flow'
        },
        data: {
          type: 'object',
          description: 'New flow data/configuration'
        },
        folder_id: {
          type: 'string',
          description: 'New folder ID'
        }
      },
      required: ['flow_id']
    }
  },
  {
    name: 'delete_flow',
    description: 'Delete a single flow by ID.',
    inputSchema: {
      type: 'object',
      properties: {
        flow_id: {
          type: 'string',
          description: 'The ID of the flow to delete'
        }
      },
      required: ['flow_id']
    }
  },
  {
    name: 'delete_flows',
    description: 'Delete multiple flows at once. Pass array of flow IDs.',
    inputSchema: {
      type: 'object',
      properties: {
        flow_ids: {
          type: 'array',
          items: { type: 'string' },
          description: 'Array of flow IDs to delete'
        }
      },
      required: ['flow_ids']
    }
  },
  {
    name: 'list_components',
    description: 'List all available Langflow components. Returns components with their types, names, and descriptions.',
    inputSchema: {
      type: 'object',
      properties: {}
    }
  },
  {
    name: 'run_flow',
    description: 'Execute a Langflow flow by ID or name. Returns execution results including outputs and session ID.',
    inputSchema: {
      type: 'object',
      properties: {
        flow_id_or_name: {
          type: 'string',
          description: 'Flow ID (UUID) or flow name'
        },
        input_request: {
          type: 'object',
          description: 'Input request configuration',
          properties: {
            input_value: {
              type: 'string',
              description: 'Input value for the flow'
            },
            output_type: {
              type: 'string',
              description: 'Expected output type'
            },
            input_type: {
              type: 'string',
              description: 'Type of input being provided'
            },
            tweaks: {
              type: 'object',
              description: 'Optional parameter tweaks'
            }
          }
        },
        stream: {
          type: 'boolean',
          description: 'Enable streaming mode (default: false)'
        }
      },
      required: ['flow_id_or_name', 'input_request']
    }
  },
  {
    name: 'trigger_webhook',
    description: 'Trigger a flow via webhook endpoint. Simplified API for webhook-based flow execution.',
    inputSchema: {
      type: 'object',
      properties: {
        flow_id_or_name: {
          type: 'string',
          description: 'Flow ID (UUID) or flow name'
        },
        input_request: {
          type: 'object',
          description: 'Simplified input request',
          properties: {
            input_value: {
              type: 'string',
              description: 'Input value for the flow'
            },
            tweaks: {
              type: 'object',
              description: 'Optional parameter tweaks'
            }
          }
        }
      },
      required: ['flow_id_or_name', 'input_request']
    }
  },
  {
    name: 'upload_flow',
    description: 'Upload a flow from file data. Provide an object with name, content (stringified JSON), and optional type fields.',
    inputSchema: {
      type: 'object',
      properties: {
        file: {
          type: 'object',
          description: 'File object with fields: name (string), content (string, max 10MB), type (optional string)',
          properties: {
            name: {
              type: 'string',
              description: 'The filename of the flow export (e.g., flow.json)'
            },
            content: {
              type: 'string',
              description: 'The contents of the flow export as a JSON string'
            },
            type: {
              type: 'string',
              description: 'Optional MIME type of the file (e.g., application/json)'
            }
          },
          required: ['name', 'content']
        }
      },
      required: ['file']
    }
  },
  {
    name: 'download_flows',
    description: 'Download flows as JSON export. Exports multiple flows for backup or transfer.',
    inputSchema: {
      type: 'object',
      properties: {
        flow_ids: {
          type: 'array',
          items: { type: 'string' },
          description: 'Array of flow IDs (UUIDs) to download'
        }
      },
      required: ['flow_ids']
    }
  },
  {
    name: 'get_basic_examples',
    description: 'Get basic example flows. Returns a list of pre-built example flows for learning and templates.',
    inputSchema: {
      type: 'object',
      properties: {}
    }
  },
  {
    name: 'list_folders',
    description: 'List all folders with optional pagination. Folders help organize flows hierarchically.',
    inputSchema: {
      type: 'object',
      properties: {
        page: {
          type: 'number',
          description: 'Page number for pagination (default: 1)'
        },
        size: {
          type: 'number',
          description: 'Number of items per page (default: 50, max: 100)'
        }
      }
    }
  },
  {
    name: 'create_folder',
    description: 'Create a new folder to organize flows. Supports nested folder structure with parent_id.',
    inputSchema: {
      type: 'object',
      properties: {
        name: {
          type: 'string',
          description: 'Folder name (required)'
        },
        description: {
          type: 'string',
          description: 'Optional folder description'
        },
        parent_id: {
          type: 'string',
          description: 'Optional parent folder ID (UUID) for nested folders'
        }
      },
      required: ['name']
    }
  },
  {
    name: 'get_folder',
    description: 'Get details of a specific folder by ID.',
    inputSchema: {
      type: 'object',
      properties: {
        folder_id: {
          type: 'string',
          description: 'Folder ID (UUID)'
        }
      },
      required: ['folder_id']
    }
  },
  {
    name: 'update_folder',
    description: 'Update an existing folder. Can change name, description, or move to different parent.',
    inputSchema: {
      type: 'object',
      properties: {
        folder_id: {
          type: 'string',
          description: 'Folder ID (UUID) to update'
        },
        name: {
          type: 'string',
          description: 'New folder name'
        },
        description: {
          type: 'string',
          description: 'New folder description'
        },
        parent_id: {
          type: 'string',
          description: 'New parent folder ID (UUID)'
        }
      },
      required: ['folder_id']
    }
  },
  {
    name: 'delete_folder',
    description: 'Delete a folder by ID. Warning: This may affect flows in the folder.',
    inputSchema: {
      type: 'object',
      properties: {
        folder_id: {
          type: 'string',
          description: 'Folder ID (UUID) to delete'
        }
      },
      required: ['folder_id']
    }
  },
  {
    name: 'list_projects',
    description: 'List all projects with optional pagination. Projects group related flows and resources.',
    inputSchema: {
      type: 'object',
      properties: {
        page: {
          type: 'number',
          description: 'Page number for pagination (default: 1)'
        },
        size: {
          type: 'number',
          description: 'Number of items per page (default: 50, max: 100)'
        }
      }
    }
  },
  {
    name: 'create_project',
    description: 'Create a new project. Projects help organize and group related flows.',
    inputSchema: {
      type: 'object',
      properties: {
        name: {
          type: 'string',
          description: 'Project name (required)'
        },
        description: {
          type: 'string',
          description: 'Optional project description'
        }
      },
      required: ['name']
    }
  },
  {
    name: 'get_project',
    description: 'Get details of a specific project by ID.',
    inputSchema: {
      type: 'object',
      properties: {
        project_id: {
          type: 'string',
          description: 'Project ID (UUID)'
        }
      },
      required: ['project_id']
    }
  },
  {
    name: 'update_project',
    description: 'Update an existing project. Can change name or description.',
    inputSchema: {
      type: 'object',
      properties: {
        project_id: {
          type: 'string',
          description: 'Project ID (UUID) to update'
        },
        name: {
          type: 'string',
          description: 'New project name'
        },
        description: {
          type: 'string',
          description: 'New project description'
        }
      },
      required: ['project_id']
    }
  },
  {
    name: 'delete_project',
    description: 'Delete a project by ID. Warning: This may affect flows in the project.',
    inputSchema: {
      type: 'object',
      properties: {
        project_id: {
          type: 'string',
          description: 'Project ID (UUID) to delete'
        }
      },
      required: ['project_id']
    }
  },
  {
    name: 'upload_project',
    description: 'Upload a project from file data. Provide an object with name, content (stringified JSON), and optional type fields.',
    inputSchema: {
      type: 'object',
      properties: {
        file: {
          type: 'object',
          description: 'File object with fields: name (string), content (string, max 10MB), type (optional string)',
          properties: {
            name: {
              type: 'string',
              description: 'File name (e.g., project.json)'
            },
            content: {
              type: 'string',
              description: 'File content as stringified JSON'
            },
            type: {
              type: 'string',
              description: 'Optional MIME type (e.g., application/json)'
            }
          },
          required: ['name', 'content']
        }
      },
      required: ['file']
    }
  },
  {
    name: 'download_project',
    description: 'Download a project as JSON export. Exports project for backup or transfer.',
    inputSchema: {
      type: 'object',
      properties: {
        project_id: {
          type: 'string',
          description: 'Project ID (UUID) to download'
        }
      },
      required: ['project_id']
    }
  },
  {
    name: 'list_variables',
    description: 'List all global variables. Variables store reusable values across flows.',
    inputSchema: {
      type: 'object',
      properties: {}
    }
  },
  {
    name: 'create_variable',
    description: 'Create a new global variable. Variables can be referenced in multiple flows.',
    inputSchema: {
      type: 'object',
      properties: {
        name: {
          type: 'string',
          description: 'Variable name (required)'
        },
        value: {
          type: 'string',
          description: 'Variable value (required)'
        },
        type: {
          type: 'string',
          description: 'Optional variable type'
        },
        default_fields: {
          type: 'array',
          items: { type: 'string' },
          description: 'Optional default fields'
        }
      },
      required: ['name', 'value']
    }
  },
  {
    name: 'update_variable',
    description: 'Update an existing variable. Can change name, value, or type.',
    inputSchema: {
      type: 'object',
      properties: {
        variable_id: {
          type: 'string',
          description: 'Variable ID (UUID) to update'
        },
        name: {
          type: 'string',
          description: 'New variable name'
        },
        value: {
          type: 'string',
          description: 'New variable value'
        },
        type: {
          type: 'string',
          description: 'New variable type'
        }
      },
      required: ['variable_id']
    }
  },
  {
    name: 'delete_variable',
    description: 'Delete a variable by ID. Warning: Flows using this variable may be affected.',
    inputSchema: {
      type: 'object',
      properties: {
        variable_id: {
          type: 'string',
          description: 'Variable ID (UUID) to delete'
        }
      },
      required: ['variable_id']
    }
  },
  {
    name: 'build_flow',
    description: 'Build/compile a flow and return job_id for polling build status. Critical for programmatic flow building and execution.',
    inputSchema: {
      type: 'object',
      properties: {
        flow_id: {
          type: 'string',
          description: 'Flow ID (UUID) to build'
        },
        inputs: {
          type: 'object',
          description: 'Optional input values for the flow'
        },
        data: {
          type: 'object',
          description: 'Optional flow data/configuration'
        },
        files: {
          type: 'array',
          items: { type: 'string' },
          description: 'Optional array of file paths'
        },
        stop_component_id: {
          type: 'string',
          description: 'Optional component ID to stop at'
        },
        start_component_id: {
          type: 'string',
          description: 'Optional component ID to start from'
        },
        log_builds: {
          type: 'boolean',
          description: 'Whether to log build process (default: true)'
        },
        flow_name: {
          type: 'string',
          description: 'Optional flow name override'
        },
        event_delivery: {
          type: 'string',
          description: 'Event delivery mode: "polling", "streaming", or "direct" (default: "polling")'
        }
      },
      required: ['flow_id']
    }
  },
  {
    name: 'get_build_status',
    description: 'Get build status and events for a specific build job. Use polling mode to check async build progress.',
    inputSchema: {
      type: 'object',
      properties: {
        job_id: {
          type: 'string',
          description: 'Build job ID returned from build_flow'
        },
        event_delivery: {
          type: 'string',
          description: 'Event delivery mode: "polling", "streaming", or "direct" (default: "polling")'
        }
      },
      required: ['job_id']
    }
  },
  {
    name: 'cancel_build',
    description: 'Cancel a running build job. Stops the build process for the specified job.',
    inputSchema: {
      type: 'object',
      properties: {
        job_id: {
          type: 'string',
          description: 'Build job ID to cancel'
        }
      },
      required: ['job_id']
    }
  },
  {
    name: 'list_knowledge_bases',
    description: 'List all available knowledge bases. Essential for RAG workflows and document management.',
    inputSchema: {
      type: 'object',
      properties: {}
    }
  },
  {
    name: 'get_knowledge_base',
    description: 'Get detailed information about a specific knowledge base by name.',
    inputSchema: {
      type: 'object',
      properties: {
        kb_name: {
          type: 'string',
          description: 'Knowledge base name/ID'
        }
      },
      required: ['kb_name']
    }
  },
  {
    name: 'delete_knowledge_base',
    description: 'Delete a specific knowledge base by name. Warning: This will permanently remove the knowledge base.',
    inputSchema: {
      type: 'object',
      properties: {
        kb_name: {
          type: 'string',
          description: 'Knowledge base name/ID to delete'
        }
      },
      required: ['kb_name']
    }
  },
  {
    name: 'bulk_delete_knowledge_bases',
    description: 'Delete multiple knowledge bases at once. Useful for batch cleanup operations.',
    inputSchema: {
      type: 'object',
      properties: {
        kb_names: {
          type: 'array',
          items: { type: 'string' },
          description: 'Array of knowledge base names to delete'
        }
      },
      required: ['kb_names']
    }
  }
];
