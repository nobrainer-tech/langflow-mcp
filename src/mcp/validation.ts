import { z } from 'zod';

export const CreateFlowSchema = z.object({
  name: z.string().min(1, 'Flow name is required').max(255, 'Flow name too long'),
  description: z.string().optional(),
  data: z.record(z.unknown()).optional(),
  folder_id: z.string().uuid('Invalid folder ID format').optional()
});

export const ListFlowsSchema = z.object({
  page: z.number().int().positive().optional(),
  size: z.number().int().positive().max(100, 'Page size cannot exceed 100').optional(),
  folder_id: z.string().uuid('Invalid folder ID format').optional()
});

export const GetFlowSchema = z.object({
  flow_id: z.string().uuid('Invalid flow ID format')
});

export const UpdateFlowSchema = z.object({
  flow_id: z.string().uuid('Invalid flow ID format'),
  name: z.string().min(1, 'Flow name is required').max(255, 'Flow name too long').optional(),
  description: z.string().optional(),
  data: z.record(z.unknown()).optional(),
  folder_id: z.string().uuid('Invalid folder ID format').optional()
}).refine(
  data => data.name !== undefined || data.description !== undefined || data.data !== undefined || data.folder_id !== undefined,
  { message: 'At least one field must be provided for update' }
);

export const DeleteFlowSchema = z.object({
  flow_id: z.string().uuid('Invalid flow ID format')
});

export const DeleteFlowsSchema = z.object({
  flow_ids: z.array(z.string().uuid('Invalid flow ID format')).min(1, 'At least one flow ID is required')
});

export const ListComponentsSchema = z.object({});

export type CreateFlowInput = z.infer<typeof CreateFlowSchema>;
export type ListFlowsInput = z.infer<typeof ListFlowsSchema>;
export type GetFlowInput = z.infer<typeof GetFlowSchema>;
export type UpdateFlowInput = z.infer<typeof UpdateFlowSchema>;
export type DeleteFlowInput = z.infer<typeof DeleteFlowSchema>;
export type DeleteFlowsInput = z.infer<typeof DeleteFlowsSchema>;
export type ListComponentsInput = z.infer<typeof ListComponentsSchema>;
