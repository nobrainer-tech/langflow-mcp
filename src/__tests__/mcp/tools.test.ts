import { describe, it, expect } from 'vitest';
import { langflowMCPTools } from '../../mcp/tools';

describe('MCP Tools Definitions', () => {
  describe('Tool Structure', () => {
    it('should have all required properties for each tool', () => {
      langflowMCPTools.forEach((tool, index) => {
        expect(tool, `Tool at index ${index} should have name`).toHaveProperty('name');
        expect(tool, `Tool at index ${index} should have description`).toHaveProperty('description');
        expect(tool, `Tool at index ${index} should have inputSchema`).toHaveProperty('inputSchema');

        expect(typeof tool.name, `Tool ${tool.name} name should be string`).toBe('string');
        expect(typeof tool.description, `Tool ${tool.name} description should be string`).toBe('string');
        expect(typeof tool.inputSchema, `Tool ${tool.name} inputSchema should be object`).toBe('object');
      });
    });

    it('should have non-empty names and descriptions', () => {
      langflowMCPTools.forEach(tool => {
        expect(tool.name.length, `Tool ${tool.name} should have non-empty name`).toBeGreaterThan(0);
        expect(tool.description.length, `Tool ${tool.name} should have non-empty description`).toBeGreaterThan(0);
      });
    });

    it('should use snake_case for tool names', () => {
      const snakeCasePattern = /^[a-z][a-z0-9]*(_[a-z0-9]+)*$/;

      langflowMCPTools.forEach(tool => {
        expect(tool.name, `Tool name ${tool.name} should be in snake_case`).toMatch(snakeCasePattern);
      });
    });

    it('should have unique tool names', () => {
      const toolNames = langflowMCPTools.map(t => t.name);
      const uniqueNames = new Set(toolNames);

      expect(uniqueNames.size).toBe(toolNames.length);
    });
  });

  describe('Input Schemas', () => {
    it('should have valid JSON schema structure', () => {
      langflowMCPTools.forEach(tool => {
        expect(tool.inputSchema.type).toBe('object');

        if (tool.inputSchema.properties) {
          expect(typeof tool.inputSchema.properties).toBe('object');
        }

        if (tool.inputSchema.required) {
          expect(Array.isArray(tool.inputSchema.required)).toBe(true);
        }
      });
    });

    it('should define property types when properties exist', () => {
      langflowMCPTools.forEach(tool => {
        if (tool.inputSchema.properties) {
          Object.entries(tool.inputSchema.properties).forEach(([propName, propSchema]: [string, any]) => {
            expect(propSchema, `Property ${propName} in ${tool.name} should have type`).toHaveProperty('type');
          });
        }
      });
    });

    it('should have description for each property', () => {
      langflowMCPTools.forEach(tool => {
        if (tool.inputSchema.properties) {
          Object.entries(tool.inputSchema.properties).forEach(([propName, propSchema]: [string, any]) => {
            if (!propSchema.description && !propSchema.anyOf && !propSchema.oneOf) {
              console.warn(`Property ${propName} in ${tool.name} missing description`);
            }
          });
        }
      });
    });
  });

  describe('Deprecated Tools', () => {
    const deprecatedToolNames = [
      'build_vertices',
      'get_vertex',
      'stream_vertex_build',
      'get_task_status'
    ];

    it('should mark all deprecated tools with warning emoji', () => {
      deprecatedToolNames.forEach(toolName => {
        const tool = langflowMCPTools.find(t => t.name === toolName);

        expect(tool, `Deprecated tool ${toolName} should exist`).toBeDefined();
        expect(tool?.description, `Deprecated tool ${toolName} should have warning emoji`).toContain('⚠️');
      });
    });

    it('should explicitly mention DEPRECATED in description', () => {
      deprecatedToolNames.forEach(toolName => {
        const tool = langflowMCPTools.find(t => t.name === toolName);

        expect(tool?.description.toUpperCase()).toContain('DEPRECATED');
      });
    });

    it('should suggest alternative tools', () => {
      const alternatives: Record<string, string[]> = {
        'build_vertices': ['build_flow'],
        'get_vertex': ['build_flow', 'get_flow'],
        'stream_vertex_build': ['get_build_status'],
        'get_task_status': ['get_build_status']
      };

      Object.entries(alternatives).forEach(([toolName, expectedAlternatives]) => {
        const tool = langflowMCPTools.find(t => t.name === toolName);

        expect(tool).toBeDefined();

        expectedAlternatives.forEach(alternative => {
          expect(
            tool?.description.toLowerCase(),
            `Tool ${toolName} should mention alternative ${alternative}`
          ).toContain(alternative.toLowerCase());
        });
      });
    });

    it('should have exactly 4 deprecated tools', () => {
      const deprecatedTools = langflowMCPTools.filter(tool =>
        deprecatedToolNames.includes(tool.name)
      );

      expect(deprecatedTools.length).toBe(4);
    });
  });

  describe('Tool Categories', () => {
    const expectedCategories = {
      flow: ['create_flow', 'list_flows', 'get_flow', 'update_flow', 'delete_flow'],
      folder: ['create_folder', 'list_folders', 'get_folder', 'update_folder', 'delete_folder'],
      build: ['build_flow', 'get_build_status', 'cancel_build'],
      execution: ['run_flow', 'trigger_webhook'],
      auth: ['login', 'logout', 'refresh_token'],
    };

    Object.entries(expectedCategories).forEach(([category, expectedTools]) => {
      it(`should have all ${category} management tools`, () => {
        expectedTools.forEach(toolName => {
          const tool = langflowMCPTools.find(t => t.name === toolName);
          expect(tool, `${category} tool ${toolName} should exist`).toBeDefined();
        });
      });
    });
  });

  describe('Critical Tools', () => {
    const criticalTools = [
      'create_flow',
      'run_flow',
      'build_flow',
      'list_flows',
      'get_flow',
      'update_flow',
      'delete_flow'
    ];

    it('should have all critical tools defined', () => {
      criticalTools.forEach(toolName => {
        const tool = langflowMCPTools.find(t => t.name === toolName);
        expect(tool, `Critical tool ${toolName} should exist`).toBeDefined();
      });
    });

    it('should have detailed descriptions for critical tools', () => {
      criticalTools.forEach(toolName => {
        const tool = langflowMCPTools.find(t => t.name === toolName);
        expect(tool?.description.length, `Critical tool ${toolName} should have detailed description`).toBeGreaterThan(50);
      });
    });
  });

  describe('Tool Count', () => {
    it('should have expected total number of tools', () => {
      // Expected: 90 active tools + 4 deprecated tools = 94 total
      expect(langflowMCPTools.length).toBeGreaterThanOrEqual(90);
      expect(langflowMCPTools.length).toBeLessThanOrEqual(100);
    });

    it('should maintain consistent tool count', () => {
      const currentCount = langflowMCPTools.length;
      expect(currentCount).toMatchSnapshot();
    });
  });

  describe('Documentation Quality', () => {
    it('should not have placeholder descriptions', () => {
      const placeholders = ['todo', 'tbd', 'fixme', 'xxx'];

      langflowMCPTools.forEach(tool => {
        const descLower = tool.description.toLowerCase();
        placeholders.forEach(placeholder => {
          expect(descLower, `Tool ${tool.name} should not have placeholder: ${placeholder}`).not.toContain(placeholder);
        });
      });
    });

    it('should start descriptions with capital letter or emoji', () => {
      langflowMCPTools.forEach(tool => {
        const firstChar = tool.description[0];
        const isCapital = /[A-Z⚠️]/.test(firstChar);
        expect(isCapital, `Tool ${tool.name} description should start with capital or emoji`).toBe(true);
      });
    });

    it('should have descriptions longer than just the tool name', () => {
      langflowMCPTools.forEach(tool => {
        const nameWords = tool.name.split('_').join(' ');
        expect(
          tool.description.length,
          `Tool ${tool.name} description should be more than just the name`
        ).toBeGreaterThan(nameWords.length + 10);
      });
    });
  });
});
