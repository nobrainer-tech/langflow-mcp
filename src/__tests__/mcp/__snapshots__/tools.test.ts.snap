// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MCP Tools Definitions > Tool Count > should maintain consistent tool count 1`] = `90`;
